SPEC.md

```diff
@@ -0,0 +1,334 @@
+
+  **Body**: A JSON serialized [`Failure`](#failure) object.
+
+- `201 Created`: Operation was started an will completed asynchronously.
+
+  **Headers**:
+
+  - `Content-Type: application/json`
+
+  **Body**: A JSON serialized [`OperationStarted`](#operationstarted) object.
+
+### Cancel Operation
+
+Request to cancel an operation.
+The operation may later complete as canceled or any other outcome.
+Handlers should ignore multiple cancelations of the same operation and return successfully if cancelation was already
+requested.
+
+#### Method
+
+`POST`
+
+#### Paths
+
+- `/api/v1/services/{service}/operations/{operation}/{operation_id}/cancel`
+
+#### Response Codes
+
+- `202 Accepted`: Cancelation request accepted.
+
+  **Body**: Unspecified.
+
+- `404 Not Found`: Operation ID not recognized or references deleted.
+
+  **Body**: Unspecified.
+
+### Get Operation Result
+
+Retrieve operation result.
+
+#### Method
+
+`GET`
+
+#### Paths
+
+- `/api/v1/services/{service}/operations/{operation}/{operation_id}/result`
+
+#### Query Parameters
+
+- `wait_deadline`: Optional. ISO 8601 date indicating the waiting period for a result.
+  If not specified, and the operation is still running, the request should resolve immediately with the current
+  operation status.
+
+#### Response Codes
+
+- `200 OK`: Operation completed successfully.
+
+  **Headers**:
+
+  - `Nexus-Operation-State: succeeded`
+
+  **Body**: Arbitrary data conveying the operation's result. Headers should specify content type and encoding.
+
+- `204 No Content`: Operation completed successfully with an empty result or is still running.
+
+  **Headers**:
+
+  - `Nexus-Operation-State: succeeded | running`
+
+  **Body**: Empty.
+
+- `482 Operation Failed`: Operation completed as `failed` or `canceled`.
+
+  **Headers**:
+
+  - `Content-Type: application/json`
+  - `Nexus-Operation-State: failed | canceled`
+
+  **Body**: A JSON serialized [`Failure`](#failure) object.
+
+- `408 Request Timeout`: The specified `wait_deadline` expired prior to operation completion.
+
+  The caller may re-issue this request to start a new long poll.
+
+  **Body**: Unspecified.
+
+- `404 Not Found`: Operation ID not recognized or references deleted.
+
+  **Body**: Unspecified.
+
+### Get Operation Info
+
+Retrieve operation details.
+
+#### Method
+
+`GET`
+
+#### Paths
+
+- `/api/v1/services/{service}/operations/{operation}/{operation_id}`
+
+#### Response Codes
+
+- `200 OK`: Successfully retrieved info.
+
+  **Headers**:
+
+  - `Content-Type: application/json`
+  - `Nexus-Operation-State: failed | canceled`
+
+  **Body**:
+
+  A JSON serialized [`OperationInfo`](#operationinfo) object.
+
+- `404 Not Found`: Operation ID not recognized or references deleted.
+
+## General Information on HTTP Response Codes
+
+The Nexus protocol follows standard HTTP practices:
+
+- `4xx` codes signify non-retryable framework-level errors.
+- `5xx` codes signify retryable framework-level errors.
```

Comment:
There aren't any 5xx responses in the spec.

Reply:
It's unreasonable to define internal errors in this high level spec IMO. It may be implementation specific.

Reply:
We don't need to define them in detail, but the document needs to mention that they may be returned if it's going to include this line stating that the protocol uses them with these semantics.